type of Union containing Any is incorrectly inferred after an isinstance check

`union_any.py`:

``` py
from typing import Any, Union

def func(v):
    # type: (Union[int, Any]) -> str
    if isinstance(v, int):
        s = hex(v)
    else:
        s = str(v).lower()
    return "key:" + s
```

`mypy union_any.py`:

```
union_any.py: note: In function "func":
union_any.py:8: error: No overload variant of "str" matches argument types [Union[<ERROR>, void]]
```

`mypy --py2 union_any.py`:

```
union_any.py: note: In function "func":
union_any.py:8: error: Argument 1 to "str" has incompatible type "Union[object, None]"; expected "object"
```

I discovered this when I used a class which was defined in some library which did not have stubs and I was using `--silent-imports`.

Another case where I found this error is when 2 classes inherit from `Any`:

`file1.py`:
```py
from typing import Any, Union

class B(Any): name = 'Bob'
class C(Any): name = 'Charlie'

def func(x):
    # type: (Union[B, C]) -> None
    reveal_type(x)
    if isinstance(x, B):
        reveal_type(x)
        print(x.name)
    else:
        reveal_type(x)
        print(x.name)
```

gives this output:
```
file1.py: note: In function "func":
file1.py:8: error: Revealed type is 'Union[file1.B, file1.C]'
file1.py:10: error: Revealed type is 'file1.B'
file1.py:13: error: Revealed type is 'Union[<ERROR>, <ERROR>]'
file1.py:14: error: Some element of union has no attribute "name"
```

It works fine when I replace `Any` by `object`.

--- a/mypy/subtypes.py
+++ b/mypy/subtypes.py
@@ -345,7 +345,8 @@ def restrict_subtype_away(t: Type, s: Type) -> Type:
     Currently just remove elements of a union type.
     """
     if isinstance(t, UnionType):
-        new_items = [item for item in t.items if not is_subtype(item, s)]
+        new_items = [item for item in t.items if (not is_subtype(item, s)
+                                                  or isinstance(item, AnyType))]
         return UnionType.make_union(new_items)
     else:
         return t

--- a/test-data/unit/check-unions.test
+++ b/test-data/unit/check-unions.test
@@ -40,7 +40,7 @@ from typing import Any, Union
 
 def func(v:Union[int, Any]) -> None:
     if isinstance(v, int):
-        reveal_type(v) # E: Revealed type is 'builins.int'
+        reveal_type(v) # E: Revealed type is 'builtins.int'
     else:
         reveal_type(v) # E: Revealed type is 'Any'
 [builtins fixtures/isinstance.pyi]

--- a/test-data/unit/check-unions.test
+++ b/test-data/unit/check-unions.test
@@ -35,6 +35,17 @@ def f(x: Union[int, str]) -> None:
 [out]
 main: note: In function "f":
 
+[case testUnionAnyIsInstance]
+from typing import Any, Union
+
+def func(v:Union[int, Any]) -> None:
+    if isinstance(v, int):
+        reveal_type(v) # E: Revealed type is 'builins.int'
+    else:
+        reveal_type(v) # E: Revealed type is 'Any'
+[builtins fixtures/isinstance.pyi]
+[out]
+main: note: In function "func":
 
 [case testUnionAttributeAccess]
 from typing import Union

